omnigibson/object_states/attached_to.py

```diff
@@ -0,0 +1,293 @@
+                return False
+
+        # Attempt to detach
+        else:
+            if self.parent == other:
+                self._detach()
+
+                # Wake up objects so that passive forces like gravity can be applied.
+                self.obj.wake()
+                other.wake()
+            return True
+
+    def _get_value(self, other):
+        # Simply return if the current parent matches other
+        return other == self.parent
+
+    def _find_attachment_links(self,
+                               other,
+                               bypass_alignment_checking=False,
+                               pos_thresh=m.DEFAULT_POSITION_THRESHOLD,
+                               orn_thresh=m.DEFAULT_ORIENTATION_THRESHOLD):
+        """
+        Args:
+            other (DatasetObject): parent object to find potential attachment links.
+            bypass_alignment_checking (bool): whether to bypass alignment checking when finding attachment links.
+            pos_thresh (float): position difference threshold to activate attachment, in meters.
+            orn_thresh (float): orientation difference threshold to activate attachment, in radians.
+
+        Returns:
+            RigidPrim or None: link belonging to @self.obj that should be aligned to that corresponding link of @other
+            RigidPrim or None: the corresponding link of @other
+        """
+        parent_candidates = self._get_parent_candidates(other)
+        if not parent_candidates:
+            return None, None
+
+        for child_link_name, parent_link_names in parent_candidates.items():
+            child_link = self.links[child_link_name]
+            for parent_link_name in parent_link_names:
+                parent_link = other.states[AttachedTo].links[parent_link_name]
+                if other.states[AttachedTo].children[parent_link_name] is None:
+                    pos_diff = np.linalg.norm(child_link.get_position() - parent_link.get_position())
+                    orn_diff = T.get_orientation_diff_in_radian(child_link.get_orientation(), parent_link.get_orientation())
+                    if bypass_alignment_checking or (pos_diff < pos_thresh and orn_diff < orn_thresh):
+                        return child_link, parent_link
+
+        return None, None
+
+    def _get_parent_candidates(self, other):
+        """
+        Helper function to return the parent link candidates for @other
+
+        Returns:
+            Dict[str, str] or None: mapping from the male meta link names of self.obj to the correspounding female meta
+            link names of @other. Returns None if @other does not have the AttachedTo state.
+        """
+        if AttachedTo not in other.states:
+            return None
+
+        if other not in self.parent_link_candidates:
+            parent_link_names = defaultdict(set)
+            for child_link_name, child_link in self.links.items():
+                child_category = child_link_name.split("_")[1]
```

Comment:
This seems a bit brittle, but I can understand that this is probably the most efficient way to infer critical info about the attachment links.

Can we add explicit comments here, as to the expected format of the links? From what I gather it sounds like it should be something like `attachment_<CATEGORY>_<M/F>`

Reply:
maybe it would be better to have the M/F separate, so it's `attachment_<CATEGORY>_<M/F>`, and so we can split the information more cleanly:

```python
child_category, child_type = child_link_name.split("_")[-2:]
if child_type == "F":
    continue
else:
    assert child_type == "M"
    parent_category_type = f"{child_category}_{child_type}"
    ...
```

Reply:
also, just realized, how will this work wrt multiple instances of the same type of attachment link? We should probably sit down with @cem and confirm the most general metalink structure, and maybe even make a a general purpose API that will programmatically decompose a given metalink name in a standardized way (e.g.: `get_metalink_info()` or something)

Reply:
Discussed in person. We will address this in the next PR that properly allows object states to use more than one meta link per type.